### PR TITLE
I331 csv from oai

### PR DESCRIPTION
**Pull request by DeonFranklin**
This PR contains the file csv_from_oai.rb in which a script was created to take an OAI feed and extracts the data for exporting to a CSV file. 

Ref. #331 

image

### CODE REVIEW

File: bin/csv_from_oai.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+# OAI client setup
+client = OAI::Client.new(
+  "http://oai.adventistdigitallibrary.org/OAI-script",
+  headers: { from: "jeremy@scientist.com" },
```
Comment (aprilrieger) [inline bin/csv_from_oai.rb:6]: Does this need to be dynamic, or will Jeremy's email forever be in the code base?

Comment (jeremyf) [reply bin/csv_from_oai.rb:6]: Job security?!? 😁

File: bin/csv_from_oai.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+}
+CSV.open('csv_from_oai.csv', 'wb') do |csv|
+  # Write the headers to the CSV file
+  csv << ['aark_id', 'identifier', 'creator',
+          'title', 'resource_type', 'date_created',
+          'language', 'extent', 'source',
+          'geocode', 'place_of_publication', 'publisher',
+          'rights_statement', 'subject', 'related_url',
+          'thumbnail_url', 'work_type']
+  sets = ["adl:thesis"]
```
Comment (jeremyf) [inline bin/csv_from_oai.rb:21]: Why are we only querying the "adl:thesis" set?

File: bin/csv_from_oai.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      thumbnail_url = r.metadata.children.first.find('thumbnail_url').first.content
+      work_type = r.metadata.children.first.find('work_type').first.content
+      # volume_number = r.metadata.children.first.find('volume_number').first.content
+      # issue_number = r.metadata.children.first.find('issue_number').first.content
+
+      # @jeremyf For the ones that are commented out, I kept getting undefined method for find or content.
+      # I am however getting the results I want on the output csv.
+
+      # # Write the data to the CSV file
+      csv << [aark_id, identifier, creator,
```
Comment (jeremyf) [inline bin/csv_from_oai.rb:58]: I was wanting one row per thumbnail_url and or related_url.  Also, I forgot to mention that there are potentially multiple files in each of those elements.  There is a splitter character that we use.

I was also wanting you to assign a derivative_type based on the detail logic.

**Review by jeremyf — CHANGES_REQUESTED** [review]

File: bin/csv_from_oai.rb (hunk 2 above)
Comment (DeonFranklin) [reply bin/csv_from_oai.rb]: fixed

File: bin/csv_from_oai.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  headers: { from: "jeremy@scientist.com" },
+  parser: 'libxml'
+)
+
+opts = {
+  metadata_prefix: "oai_adl"
+}
+CSV.open('csv_from_oai.csv', 'wb') do |csv|
+  # Write the headers to the CSV file
+  csv << ['aark_id', 'identifier', 'creator',
```
[the remark this answers is not in the document]
Comment (DeonFranklin) [reply bin/csv_from_oai.rb]: updated and fixed

File: bin/csv_from_oai.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  # "adl:article",
+  # "adl:image",
+  # "adl:book",
+  # "adl:other"]
+
+  # Create a new CSV file
+  sets.each do |set|
+    records = client.list_records(opts.merge(set: set))
+    records.each_with_index do |r, _i|
+      # Extract the data from the OAI record
```
[the remark this answers is not in the document]
Comment (DeonFranklin) [reply bin/csv_from_oai.rb]: this code was revised

File: bin/csv_from_oai.rb (hunk 3 above)
Comment (DeonFranklin) [reply bin/csv_from_oai.rb]: done

File: bin/csv_from_oai.rb (hunk 1 above)
Comment (DeonFranklin) [reply bin/csv_from_oai.rb]: email was made dynamic

File: bin/csv_from_oai.rb [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    records = client.list_records(opts.merge(set: set))
+    records.each do |r|
+      identifier = r.header.identifier
+      thumbnail_urls = r.metadata.first.find('thumbnail_url').map(&:content)
+      related_urls = r.metadata.first.find('related_url').map(&:content)
+
+      thumbnail_urls.each do |thumbnail_url|
+        # Split thumbnail_url and extract file type
+        if thumbnail_url.include?(';')
+          thumbnail_url, query_string = thumbnail_url.split(';')
```
Comment (jeremyf) [inline bin/csv_from_oai.rb:43]: Talk to me about this line.  Why the variable name `query_string`?  The query string is what follows a `?` in a URL.  So I'm confused why you're splitting on a `;` and calling one of the variables a query string.

File: bin/csv_from_oai.rb [hunk 7]
```diff
@@ -0,0 +1,80 @@
+require 'csv'
+require 'oai'
+require 'byebug'
```
Comment (jeremyf) [inline bin/csv_from_oai.rb:3]: Before we merge this, we'll want to remove the `require "byebug"; it is not best practices to keep a require of that library in code that might run in production.

File: bin/csv_from_oai.rb [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    "adl:issue",
+    "adl:article",
+    "adl:image",
+    "adl:book",
+    "adl:other"
+  ]
+
+  sets.each do |set|
+    records = client.list_records(opts.merge(set: set))
+    records.each do |r|
```
Comment (jeremyf) [inline bin/csv_from_oai.rb:35]: Reading Notes: the current implementation, which works great for testing, will only get the first page of each set.

This should have `records.full.each do |record|`

I'm wondering, how might we structure our code to allow us to choose if we want to run the `records.full.each` or the `records.each`?  For most tests, we probably don't want to run `records.full.each` but when we generate the actual CSV, we'll want that.

File: bin/csv_from_oai.rb [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+      identifier = r.header.identifier
+      thumbnail_urls = r.metadata.first.find('thumbnail_url').map(&:content)
```
Comment (jeremyf) [inline bin/csv_from_oai.rb:37]: What happens if `thumbnail_url` is `nil`?

File: bin/csv_from_oai.rb [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+      related_urls = r.metadata.first.find('related_url').map(&:content)
```
Comment (jeremyf) [inline bin/csv_from_oai.rb:38]: What happens if `related_url` is `nil`?

File: bin/csv_from_oai.rb [hunk 11]
```diff
@@ -0,0 +1,80 @@
+require 'csv'
+require 'oai'
```
Comment (jeremyf) [inline bin/csv_from_oai.rb:2]: Reminder fo later; we'll need to ensure that we're loading the `./lib/oai/client_decorator.rb` as it contains the necessary overrides to account for the unique OAI implementation of the existing server.

We'll want to add:

```ruby
require_relative "../lib/oai/client_decorator"
```

File: bin/csv_from_oai.rb (hunk 10 above)
Comment (jeremyf) [inline bin/csv_from_oai.rb:38]: Also, I see you're calling `r.metadata.first` twice, what is something you could do to reduce those calls?

File: bin/csv_from_oai.rb [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        else
+          derivative_type = "unknown"
+        end
+
+        # Write related URL to CSV
+        csv << [set, identifier, derivative_type, related_url]
+      end
+    end
+  end
+end
```
[the remark this answers is not in the document]
Comment (DeonFranklin) [reply bin/csv_from_oai.rb:80]: image

File: bin/csv_from_oai.rb (hunk 7 above)
Comment (DeonFranklin) [reply bin/csv_from_oai.rb:3]: Removed

File: bin/csv_from_oai.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (jeremyf) [reply bin/csv_from_oai.rb:80]: I'm surprised there's no text files in those.

File: bin/csv_from_oai.rb (hunk 6 above)
Comment (DeonFranklin) [reply bin/csv_from_oai.rb:43]: I did this as a means of splitting the files from the URLs based on the fact two remote files were writing to the same line.

Comment (jeremyf) [reply bin/csv_from_oai.rb:43]: What were you seeing in the thumbnail_url element that had the `;`?  My suspicion is that you didn't have a query string to the right of the first `;` but instead that you had multiple URLs.  As in multiple thumbnails.

Note this also applies to the other section that has `query_string`

File: bin/csv_from_oai.rb (hunk 8 above)
Comment (DeonFranklin) [reply bin/csv_from_oai.rb:35]: Updated

File: bin/csv_from_oai.rb [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    records.each_with_index do |r|
+    # For all records, comment out previous line and comment in the following.
+    # records.full.each_with_index do |r|
+      identifier = r.header.identifier
+      thumbnail_urls = r.metadata.first.find('thumbnail_url').map(&:content)
+      related_urls = r.metadata.first.find('related_url').map(&:content)
+
+      thumbnail_urls.each do |thumbnail_url|
+        derivative_type = "thumbnail"
+        if thumbnail_file_type = thumbnail_url.split(/[.;]/).last
```
Comment (jeremyf) [inline bin/csv_from_oai.rb:45]: What is the text of the `related_url` at this point?  

I see that you are splitting on the regular expression `/[.;]/` which is very similar to splitting on `;`.  But somewhat more confusing is that you switched from using the `first` value from the split and are now using the `last` value.  Why the switch from `first` to `last`?

File: bin/csv_from_oai.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (jeremyf) [reply bin/csv_from_oai.rb:65]: With the change, what does the CSV look like?

File: bin/csv_from_oai.rb (hunk 6 above)
Comment (DeonFranklin) [reply bin/csv_from_oai.rb]: image
I have been trying to figure out a solution for splitting " pdf ; txt "

File: bin/csv_from_oai.rb (hunk 11 above)
Comment (DeonFranklin) [reply bin/csv_from_oai.rb]: Added to code

File: bin/csv_from_oai.rb (hunk 9 above)
Comment (DeonFranklin) [reply bin/csv_from_oai.rb:40]: Added conditional to handle nil errors.

File: bin/csv_from_oai.rb (hunk 10 above)
Comment (DeonFranklin) [reply bin/csv_from_oai.rb:41]: Added conditional to handle nil errors.

File: bin/csv_from_oai.rb [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    # records.full.each_with_index do |r|
+      identifier = r.header.identifier
+      
+      thumbnail_urls = r.metadata.first.find('thumbnail_url').map(&:content)
+      related_urls = r.metadata.first.find('related_url').map(&:content)
+
+      thumbnail_urls.each do |thumbnail_url|
+        thumbnail_urls.nil? ? "This value returned as nil." : thumbnail_urls
+        derivative_type = "thumbnail"
+        if thumbnail_file_type = thumbnail_url.split(/[.;]/).each
```
Comment (jeremyf) [inline bin/csv_from_oai.rb:47]: I don't think this is behaving how you think it is behaving.  The `.each` without a block will return an Enumerator object, which means that based on the following logic you'll always mark this as a `"thumbnail"`.

Let's break this down further:

In the above what is `thumbnail_urls`, give me a concrete example for the first record.  Then what what is the first `thumbnail_url` that we yield for the each block.  Again a concrete example.

File: bin/csv_from_oai.rb [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+          derivative_type = thumbnail_file_type == 'pdf' && thumbnail_url.include?(".ARCHIVAL.pdf") ? 'original' : 'thumbnail'
+        end
+        csv << [set, identifier, derivative_type, thumbnail_url]
+      end
+
+      related_urls.each do |related_url|
+        related_urls.nil? ? "This value returned as nil." : related_urls
+        derivative_type = case related_file_type = related_url.split(/[.;]/).each
```
Comment (jeremyf) [inline bin/csv_from_oai.rb:55]: This will almost certainly always return `"unknown"`. And I don't think it's ever calling 57 and 59.

**Pull request closed without merging**